pubsublite/internal/pubsubabs/publisher.go

```diff
@@ -0,0 +1,132 @@
+
+// ErrorStop defines the Stop and Error methods.
+type ErrorStop interface {
+	// Stop sends all remaining published messages and closes publish streams.
+	// Returns once all outstanding messages have been sent or have failed to be
+	// sent. Stop should be called when the client is no longer required.
+	Error() error
+	// Error returns the error that caused the publisher client to terminate. The
+	// error returned here may contain more context than PublishResult errors. The
+	// return value may be nil if Stop() was called.
+	Stop()
+}
+
+// PublisherClient represents a PubSub PublisherClient.
+type PublisherClient interface {
+	ErrorStop
+	Publish(ctx context.Context, msg *pubsub.Message) *pubsub.PublishResult
+}
+
+// Publisher defines an abstracted interface that can be used for mocking or
+// testing purposes. It wraps a Publisher client.
+type Publisher interface {
+	ErrorStop
+	// Publish publishes `msg` to the topic asynchronously. Messages are batched and
+	// sent according to the client's PublishSettings. Publish never blocks.
+	//
+	// Publish returns a non-nil PublishResult which will be ready when the
+	// message has been sent (or has failed to be sent) to the server. Retry-able
+	// errors are automatically handled. If a PublishResult returns an error, this
+	// indicates that the publisher client encountered a fatal error and can no
+	// longer be used. Fatal errors should be manually inspected and the cause
+	// resolved. A new publisher client instance must be created to republish failed
+	// messages.
+	//
+	// Once Stop() has been called or the publisher client has failed permanently
+	// due to an error, future calls to Publish will immediately return a
+	// PublishResult with error ErrPublisherStopped.
+	//
+	// Error() returns the error that caused the publisher client to terminate and
+	// may contain more context than the error returned by PublishResult.
+	Publish(ctx context.Context, msg *pubsub.Message) PublishResult
+}
+
+// Wrap returns an abstracted Publisher from a PublisherClient.
+func Wrap(c PublisherClient) Publisher { return &pubSubClient{c} }
+
+type pubSubClient struct {
+	PublisherClient
+}
+
+func (p *pubSubClient) Publish(ctx context.Context, msg *pubsub.Message) PublishResult {
+	return p.PublisherClient.Publish(ctx, msg)
+}
+
+// PublishResult abstracts the pubsub.PublishResult type.
+type PublishResult interface {
+	// Ready returns a channel that is closed when the result is ready.
+	// When the Ready channel is closed, Get is guaranteed not to block.
+	Get(ctx context.Context) (serverID string, err error)
+	// Get returns the server-generated message ID and/or error result of a Publish call.
+	// Get blocks until the Publish call completes or the context is done.
+	Ready() <-chan struct{}
+}
+
+// IPublishResult is a copy from the internal pubsub packgage. All the code
+// below has been copied from "cloud.google.com/go/internal/pubsub"
+// Copyright 2020 Google LLC
+type IPublishResult struct {
```

Comment:
It seems to be required only for a single test, what do you think about moving it to the test file?

Reply:
Actually, it's not done in this PR but we'll need to use this type in the `pubsublite` package too, not only for that telemetry test. We could copy the type to each package `_test.go` files, or leave it here. Do you have any preferences?

Reply:
I see, I think it should be okay to leave it here then.

Reply:
I'll resolve this, then.